curvefs/src/mds/topology/topology.cpp

```diff
@@ -1257,8 +1281,152 @@ TopoStatusCode TopologyImpl::ChooseNewMetaServerForCopyset(
     return TopoStatusCode::TOPO_OK;
 }
 
-TopoStatusCode TopologyImpl::ChooseAvailableMetaServers(
-    std::set<MetaServerIdType> *metaServers, PoolIdType *poolId) {
+TopoStatusCode TopologyImpl::GenCopysetAddrRandomBatch(
+    int needCreateNum, std::list<CopysetCreateInfo>* copysetList) {
+    ReadLockGuard rlockPool(poolMutex_);
+    ReadLockGuard rlockMetaserver(metaServerMutex_);
+
+    if (needCreateNum == 0) {
+        LOG(WARNING) << "GenCopysetAddrRandomBatch want to create 0 copyset"
+                     << "return TOPO_OK";
+        return TopoStatusCode::TOPO_OK;
+    }
+
+    if (needCreateNum < 0) {
+        LOG(ERROR) << "GenCopysetAddrRandomBatch needCreateNum = "
+                   << needCreateNum << ", the num is error";
+        return TopoStatusCode::TOPO_INTERNAL_ERROR;
+    }
+
+    LOG(INFO) << "GenCopysetAddrRandomBatch needCreateNum = "
+              << needCreateNum << " begin";
+
+    // 1. genarate replicaMap
+    std::map<PoolIdType, uint16_t> replicaMap;
+    for (const auto &it : poolMap_) {
+        replicaMap.emplace(it.first, it.second.GetReplicaNum());
+    }
+
+    // 2. get all meet the conditions metaserver
+    std::vector<const MetaServer *> vec;
+    for (const auto &it : metaServerMap_) {
+        if (it.second.GetOnlineState() == OnlineState::ONLINE
+         && it.second.GetMetaServerSpace().IsMetaserverResourceAvailable()) {
```

Comment:
In addition to considering memory and disk utilization, the number of copies also needs to be considered. 
If there are too many replication groups, no matter how the number of balanced leaders on this metaserver is large, it should become a bottleneck.